crystal/src/y3storage/subvol_specification.cr

```diff
@@ -0,0 +1,212 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+module Y2Storage
+  # Helper class to represent a subvolume specification as defined
+  # in control.xml
+  #
+  class SubvolSpecification
+    property path : String
+    property copy_on_write : Bool
+    property archs : Array(String)
+
+    COW_SUBVOL_PATHS = [
+      "home",
+      "opt",
+      "srv",
+      "tmp",
+      "usr/local",
+      "var/cache",
+      "var/crash",
+      "var/lib/machines",
+      "var/lib/mailman",
+      "var/lib/named",
+      "var/log",
+      "var/opt",
+      "var/spool",
+      "var/tmp",
+      "boot/grub2/i386-pc",
+      "boot/grub2/x86_64-efi",
+    ]
+
+    # No Copy On Write for SQL databases and libvirt virtual disks to
+    # minimize performance impact
+    NO_COW_SUBVOL_PATHS = [
+      "var/lib/libvirt/images",
+      "var/lib/mariadb",
+      "var/lib/mysql",
+      "var/lib/pgsql"
+    ]
+
+    # Subvolumes, from the lists above, that contain architecture modifiers
+    SUBVOL_ARCHS = {
+      "boot/grub2/i386-pc"          => ["i386", "x86_64"],
+      "boot/grub2/x86_64-efi"       => ["x86_64"],
+    }
+
+    def initialize(path, copy_on_write = true, archs = [] of String)
+      @path = path
+      @copy_on_write = copy_on_write
+      @archs = archs
+    end
+
+    def to_s
+      text = "SubvolSpecification #{@path}"
+      text += " (NoCOW)" unless @copy_on_write
+      text
+    end
+
+    def arch_specific?
+      !archs.empty?
+    end
+
+    # Comparison operator for sorting
+    #
+    def <=>(other)
+      path <=> other.path
+    end
+
+    # Check if this subvolume should be used for the current architecture.
+    # A subvolume is used if its archs contain the current arch.
+    # It is not used if its archs contain the current arch negated
+    # (e.g. "!ppc").
+    #
+    # @return [Boolean] true if this subvolume matches the current architecture
+    #
+    def current_arch?
+      true
+    end
+
+    # Check if this subvolume should be used for an architecture.
+    #
+    # If a block is given, the block is called as the matcher with the
+    # architecture to be tested as its argument.
+    #
+    # If no block is given (and only then), the 'target_arch' parameter is
+    # used to check against.
+    #
+    # @return [Boolean] true if this subvolume matches
+    #
+    def matches_arch?(target_arch = nil, &block)
+      return true unless arch_specific?
+
+      use_subvol = false
+      archs.each do |a|
+        arch = a.dup
+        negate = arch.start_with?("!")
+        arch[0] = "" if negate # remove leading "!"
+        match = block_given? ? block.call(arch) : arch == target_arch
+        if match && negate
+          log.info("Not using #{self} for explicitly excluded arch #{arch}")
```

Comment:
Oops, I guess these calls to `log.info` should not be here. Not here and not in the Ruby class. Looks like I forgot to remove them.

Reply:
OK, will remove.